async_client.c

```diff
@@ -0,0 +1,242 @@
+          return (NULL);
+        }
+      print_error ("[acl sender] After task send\n");
+    }
+
+  return (NULL);
+}
+
+bool
+run_async_client (config_t *config)
+{
+  async_client_context_t ctx;
+  memset (&ctx, 0, sizeof (ctx));
+
+  if (thread_pool_init (&ctx.thread_pool) == S_FAILURE)
+    {
+      print_error ("Could not initialize thread pool\n");
+      return (false);
+    }
+  if (queue_init (&ctx.task_queue, sizeof (task_t)) != QS_SUCCESS)
+    {
+      print_error ("Could not initialize task queue\n");
+      return (false);
+    }
+  if (queue_init (&ctx.result_queue, sizeof (result_t)) != QS_SUCCESS)
+    {
+      print_error ("Could not initialize result queue\n");
+      return (false);
+    }
+  if (pthread_mutex_init (&ctx.mutex, NULL) != 0)
+    {
+      print_error ("Could not initialize mutex\n");
+      return (false);
+    }
+  if (pthread_cond_init (&ctx.cond_sem, NULL) != 0)
+    {
+      print_error ("Could not initialize conditional semaphore\n");
+      return (false);
+    }
+  ctx.config = config;
+  ctx.done = false;
+  ctx.config->hash = ctx.hash;
+  ctx.config->alph = ctx.alph;
+
+  ctx.socket_fd = socket (AF_INET, SOCK_STREAM, 0);
+  if (ctx.socket_fd == -1)
+    {
+      print_error ("Could not initialize client socket\n");
+      return (false);
+    }
+
+  struct sockaddr_in addr;
+  addr.sin_family = AF_INET;
+  addr.sin_addr.s_addr = inet_addr (config->addr);
+  addr.sin_port = htons (config->port);
+
+  if (connect (ctx.socket_fd, (struct sockaddr *)&addr, sizeof (addr)) == -1)
+    {
+      print_error ("Could not connect to server\n");
+      return (false);
+    }
+
+  async_client_context_t *ctx_ptr = &ctx;
+
+  thread_create (&ctx.thread_pool, task_receiver, &ctx_ptr, sizeof (ctx_ptr));
+  print_error ("[async client] Created receiver thread\n");
+  thread_create (&ctx.thread_pool, result_sender, &ctx_ptr, sizeof (ctx_ptr));
+  print_error ("[async client] Created sender thread\n");
+  create_threads (&ctx.thread_pool, config->number_of_threads, client_worker,
+                  &ctx_ptr, sizeof (ctx_ptr));
+  print_error ("[async client] Created worker threads\n");
+
+  if (pthread_mutex_lock (&ctx.mutex) != 0)
+    {
+      print_error ("Could not lock a mutex\n");
+      return (S_FAILURE);
+    }
+  pthread_cleanup_push (cleanup_mutex_unlock, &ctx.mutex);
+
+  while (!ctx.done)
+    if (pthread_cond_wait (&ctx.cond_sem, &ctx.mutex) != 0)
+      {
+        print_error ("Could not wait on a condition\n");
+        return (S_FAILURE);
+      }
+
+  pthread_cleanup_pop (!0);
+
+  print_error ("[async client] After wait\n");
+
+  if (queue_cancel (&ctx.task_queue) != QS_SUCCESS)
+    {
+      print_error ("Could not cancel task queue\n");
+      return (false);
+    }
+  if (queue_cancel (&ctx.result_queue) != QS_SUCCESS)
+    {
+      print_error ("Could not cancel result queue\n");
+      return (false);
+    }
+  if (thread_pool_cancel (&ctx.thread_pool) == S_FAILURE)
```

Comment:
We need to gracefully collect threads. queue_cancel will initiate graceful shutdown.
change thread_pool_cancel on thread_pool_join